Add fit_discrete_mc

Smit-create opened the pull request:
#678

oyamad reviewed (commented):
Thanks @Smit-create

* [ ] This part is missing: https://github.com/QuantEcon/QuantEcon.py/blob/745c0d447a95277aadbbc66c50a6dc9656bafe36/quantecon/markov/approximation.py#L371-L373
* [ ] Add tests
* [ ] `Examples` section will be helpful

There are a few points open for discussion:
* What shape should `X` be of?
* Is the name `S` of the second argument appropriate?
* Is this for approximation or estimation (which of `approximation.py` or `estimate.py` to include this code in)?

jstac commented:
Thanks for the review @oyamad .

I think this should be in estimate.py because there is no specified model to approximate.  I suggest a better name for S is grids.

I am away from my computer so I won't give further feedback.  @Smit-create , it would be great if you could coordinate with @oyamad to finish this PR.

Smit-create commented:
> There are a few points open for discussion:
> 
> * What shape should `X` be of?
> * Is the name `S` of the second argument appropriate?
> * Is this for approximation or estimation (which of `approximation.py` or `estimate.py` to include this code in)?

Thanks, @oyamad. I have addressed these questions and added the missing part. If the current diff looks good, I can start working on Examples and tests.

oyamad commented:
* Regarding the shape of `X`, I would prefer to follow the convention in `estimate_mc(X)` and `cartesian_nearest_index(X, ...)` where `X` is supposed to be a `T x n` array with `X[t]` containing the `t`-th observation. What do you think?
* For the second argument, I would follow the suggestion from @jstac (plural `grids`).

Smit-create commented:
> * where `X` is supposed to be a `T x n` array

I kept it as `n x T` because the output of `simulate_linear_model` is `n x T`, moreover, one can easily think of this as horizontally stacked individual vectors of shape `n x 1`.

> I would follow the suggestion from

Sure will change that.

jstac commented:
> I kept it as `n x T` because the output of `simulate_linear_model` is `n x T`

Both @oyamad and @Smit-create 's opinions are sensible on this point.

Another option --- combining these opinions but with a breaking change --- is to convert `linear_state_space` so that its output is $T \times n$.

This is more work but I personally prefer it, so that we have consistency across all of these functions.

oyamad reviewed (commented):
When `n=1`, it is supposed to pass for example `X = [[0.1], [0.6], [1.2], [0.2]]` and `grids = ([0., 0.5, 1.],)`. Is there any clever way to detect this case and allow passing `X = [0.1, 0.6, 1.2, 0.2]` and `grids = [0., 0.5, 1.]`?

Smit-create commented:
> Is there any clever way to detect this case

Thanks for the review. I am not sure if we should support this as we are clearly expecting the input shape of `T X n` which is independent of `n`. IMHO, it is good to keep it unchanged regarding this.

oyamad commented on quantecon/markov/estimate.py:
```diff
@@ -52,3 +53,50 @@ def _count_transition_frequencies(index_series, trans_counter):
+    >>> X = [(-0.1, 1.2), (2, 0), (0.6, 0.4), (1.0, 0.1)]
+    >>> mc = fit_discrete_mc(X, grids)
+    >>> mc.state_values
+    array([[0, 1],
+        [1, 0],
+        [2, 0]])
```
Mind the indentation:
```suggestion
    array([[0, 1],
           [1, 0],
           [2, 0]])
```

Smit-create replied:
Thanks, resolved.

oyamad commented on quantecon/markov/estimate.py:
```diff
@@ -52,3 +53,50 @@ def _count_transition_frequencies(index_series, trans_counter):
+        [1, 0],
+        [2, 0]])
+    >>> mc.P
+    array([[0., 0., 1.],
+        [0., 1., 0.],
+        [0., 1., 0.]])
```
```suggestion
    array([[0., 0., 1.],
           [0., 1., 0.],
           [0., 1., 0.]])
```

oyamad commented on quantecon/markov/estimate.py:
```diff
@@ -52,3 +53,50 @@ def _count_transition_frequencies(index_series, trans_counter):
+    ----------
+
+    X: array_like(ndim=2)
+        Time-series such that the t-th row being :math:`x_t`.
```
```suggestion
        Time-series such that the t-th row is :math:`x_t`.
```
or
```suggestion
        Time-series with the t-th row being :math:`x_t`.
```

oyamad commented on quantecon/markov/estimate.py:
```diff
@@ -52,3 +53,50 @@ def _count_transition_frequencies(index_series, trans_counter):
+        It should be of the shape T x n, where n is the number of dimensions.
+
+    grids: array_like(array_like(ndim=1))
+        Array of sorted arrays. Set of grid points in each dimension
```
```suggestion
        Array of `n` sorted arrays. Set of grid points in each dimension
```

oyamad commented on quantecon/markov/tests/test_estimate.py:
```diff
@@ -51,3 +51,34 @@ def test_mult_dim_state(self):
+    def test_order_f(self):
+        mc = fit_discrete_mc(self.X, self.grids, order='F')
+        S_expected = np.array([[1, 0], [2, 0], [0, 1], [2, 3], [3, 4]])
+        P_expected = np.array([[1., 0., 0., 0., 0.],
+                            [0., 0., 0., 1., 0.],
+                            [0., 1., 0., 0., 0.],
+                            [0., 0., 0., 0., 1.],
+                            [1., 0., 0., 0., 0.]])
```
Indentation

oyamad commented on quantecon/markov/tests/test_estimate.py:
```diff
@@ -51,3 +51,34 @@ def test_mult_dim_state(self):
+    def test_order_c(self):
+        mc = fit_discrete_mc(self.X, self.grids, order='C')
+        S_expected = np.array([[0, 1], [1, 0], [2, 0], [2, 3], [3, 4]])
+        P_expected = np.array([[0., 0., 1., 0., 0.],
+                            [0., 1., 0., 0., 0.],
+                            [0., 0., 0., 1., 0.],
+                            [0., 0., 0., 0., 1.],
+                            [0., 1., 0., 0., 0.]])
```
Indentation

oyamad reviewed (approved):
@Smit-create Thanks!

jstac commented:
Many thanks @Smit-create and @oyamad
